### PR TITLE
Inline form element path

### PR DESCRIPTION
**Pull request by timja**
See https://github.com/jenkinsci/form-element-path-plugin/pull/11#issuecomment-969148801

This inlines form-element-path so that we can change it as part of UI re-work without having to update a separate test tool and get it released. It will also enable htmlunit testing with more stable selectors which will be nice.

It is enabled by:
* default in unit tests
* system property `jenkins.formelementpath.FormElementPathPageDecorator.enabled=true
*setting `jenkins.formelementpath.FormElementPathPageDecorator.ENABLED = true` via script console.

ATH is setting system property with https://github.com/jenkinsci/acceptance-test-harness/pull/712, all ATH tests are passing with this.

Plugin is EOLed in https://github.com/jenkinsci/form-element-path-plugin/pull/12

For anyone who don't know what this is it's a plugin that provides an attribute on the DOM which is based on the 'path' to an attribute:

image

### Proposed changelog entries

* Add path to form elements giving stable selectors for UI testing.

### Proposed upgrade guidelines

N/A

### Submitter checklist

- [ ] (If applicable) Jira issue is well described
- [ ] Changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developer, depending on the change). [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
  * Fill-in the `Proposed changelog entries` section only if there are breaking changes or other changes which may require extra steps from users during the upgrade
- [ ] Appropriate autotests or explanation to why this change has no tests
- [ ] For dependency updates: links to external changelogs and, if possible, full diffs

### Desired reviewers

@mention

### Maintainer checklist

Before the changes are marked as `ready-for-merge`: 

- [ ] There are at least 2 approvals for the pull request and no outstanding requests for change
- [ ] Conversations in the pull request are over OR it is explicit that a reviewer does not block the change
- [ ] Changelog entries in the PR title and/or `Proposed changelog entries` are correct
- [ ] Proper changelog labels are set so that the changelog can be generated automatically
- [ ] If the change needs additional upgrade steps from users, `upgrade-guide-needed` label is set and there is a `Proposed upgrade guidelines` section in the PR title. ([example](https://github.com/jenkinsci/jenkins/pull/4387))
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins-ci.org/issues/?filter=12146)).

### CODE REVIEW

File: core/src/main/resources/jenkins/formelementpath/FormElementPathPageDecorator/footer.jelly
Comment (timja) [inline core/src/main/resources/jenkins/formelementpath/FormElementPathPageDecorator/footer.jelly:4]:
```suggestion
</j:jelly>

```

File: core/src/main/java/jenkins/formelementpath/FormElementPathPageDecorator.java
Comment (timja) [inline core/src/main/java/jenkins/formelementpath/FormElementPathPageDecorator.java:9]:
```suggestion
```

File: core/src/main/resources/jenkins/formelementpath/form-element-path.js
Comment (jglick) [inline core/src/main/resources/jenkins/formelementpath/form-element-path.js:26]: Probably an opportunity to clean up here.

File: core/src/main/resources/jenkins/formelementpath/form-element-path.js
Comment (jglick) [inline core/src/main/resources/jenkins/formelementpath/form-element-path.js:222]: ditto

File: core/src/main/resources/jenkins/formelementpath/form-element-path.js
Comment (jglick) [inline core/src/main/resources/jenkins/formelementpath/form-element-path.js:228]: Will this blow up somehow if the plugin is also installed?

**Review by jglick — COMMENTED** [review]
> whether it should be enabled by default or with a system property

No strong opinion. Seems like it would be simpler overall to have this enabled unconditionally, so we could inline these additions into regular controls (probably generated server-side) rather than all these tricks with `Behavior`. If we are concerned about HTML bloat and do not want it on by default, then it could be gated by a system property, defaulting to `Main.isUnitTest` perhaps.

File: core/src/main/resources/jenkins/formelementpath/form-element-path.js
Comment (timja) [reply core/src/main/resources/jenkins/formelementpath/form-element-path.js:228]: I plan to bump the baseline of the plugin and empty it out initially so this shouldn’t really matter

File: core/src/main/resources/jenkins/formelementpath/form-element-path.js
Comment (timja) [reply core/src/main/resources/jenkins/formelementpath/form-element-path.js:26]: Possibly yes I saw this but didn’t quite understand it, I’ll give it a bit of a test.

FTR I did remove some more obvious cruft from old versions

**Comment by timja** [comment]
> rather than all these tricks with Behavior

I've removed all the prototype.js code, we could inline this into `hudson-behaviour.js#buildFormTree` as well, but would rather keep it separate initially

File: core/src/main/resources/jenkins/formelementpath/form-element-path.js
Comment (timja) [inline core/src/main/resources/jenkins/formelementpath/form-element-path.js:180]: this is what I'm most worried about having it on by default, every second recalculating the paths.

I tried removing it but its needed for repeatable elements to have the index added and re-calculated.

Comment (fqueiruga) [reply core/src/main/resources/jenkins/formelementpath/form-element-path.js:180]: Jenkins already has runtime performance issues and periodic queries, I would avoid this as much as possible. 

I'd suggest maybe registering this as a layout update callback (https://github.com/jenkinsci/jenkins/blob/3e8afc53de7824a8d6fc113dff655aeb7174c55f/war/src/main/webapp/scripts/hudson-behavior.js#L2487). Most widgets already use it after adding new stuff (https://github.com/jenkinsci/jenkins/search?q=layoutUpdateCallback). Only `repeatable` is missing I think.

Comment (timja) [reply core/src/main/resources/jenkins/formelementpath/form-element-path.js:180]: it's already registered a couple of lines down, I'll take a look at adding it to repeatable to see if it removes the need for this.

Comment (fqueiruga) [reply core/src/main/resources/jenkins/formelementpath/form-element-path.js:180]: It should probably fix it for most plugins. Don't know if it's a problem that rogue plugins with custom widgets do not have this tbh

Comment (jglick) [reply core/src/main/resources/jenkins/formelementpath/form-element-path.js:180]: > Jenkins already has runtime performance issues and periodic queries, I would avoid this as much as possible.

As in https://github.com/jenkinsci/jenkins/pull/5926#pullrequestreview-809223262 you should _either_

* Add the new HTML attributes or whatever unconditionally during Jelly generation, deleting all the JavaScript hacks.
* Or disable all of this by default in production code.

Comment (timja) [reply core/src/main/resources/jenkins/formelementpath/form-element-path.js:180]: Will probably disable by default but also look to add repeatable to layout so the timer bit can be removed.

The timer bit made debugging this code a pain.

Comment (jglick) [reply core/src/main/resources/jenkins/formelementpath/form-element-path.js:180]: Maybe start with the current code copied-and-pasted from the plugin but disabled by default. That at least solves the immediate issue, we can gut the plugin, and simplify ATH to not try to load it.

Then as a follow-up look into whether the actual behavior can be simplified by inlining into `lib/form/*.jelly` and/or `hudson-behavior.js`.

**Comment by timja** [comment]
Review comments should be all addressed now.

a PR to adjust ATH:
https://github.com/jenkinsci/acceptance-test-harness/pull/712

File: core/src/main/resources/jenkins/formelementpath/form-element-path.js
Comment (timja) [reply core/src/main/resources/jenkins/formelementpath/form-element-path.js:182]: it wouldn't blow up anyway just last one to register would win

**Comment by timja** [comment]
https://github.com/jenkinsci/acceptance-test-harness/pull/712/checks?check_run_id=4291377477 shows that something here isn't working, likely it's the removal of the run every second. I'll see if there's a fix otherwise I'll restore it since this is disabled by default now

**Comment by timja** [comment]
The general case was failing because `yui-button` reworks the DOM, didn't really want to patch the library so I've just re-ran apply after 1 second.

Other case I noticed was pipeline code mirror, same workaround fixes that.

**Comment by timja** [comment]
One plugin left failing atm: `run_groovy – plugins.GroovyPluginTest`

Will see if patchable otherwise will restore the run every second hack I guess

**Comment by timja** [comment]
All is passing now in ATH, @jtnord can you take a look please?

File: core/src/main/java/jenkins/formelementpath/FormElementPathPageDecorator.java
Comment (jtnord) [inline core/src/main/java/jenkins/formelementpath/FormElementPathPageDecorator.java:13]: I wonder if we should still be able to flip this switch via the installation of a plugin?
Currently we spin up some Jenkins' instance where we do not configure system properties (or doing so is hard) and the current flow is to install the `form-element-path` plugin without the use of the paths it provides. (e.g. auto provisioning in kubernetes)

@Vlatombe knows more about this area of the framework that we use internally.

File: core/src/main/java/jenkins/formelementpath/FormElementPathPageDecorator.java
Comment (jtnord) [inline core/src/main/java/jenkins/formelementpath/FormElementPathPageDecorator.java:14]: possible future enhancement would be to also set this for `RealJenkinsRule`

File: core/src/main/resources/jenkins/formelementpath/form-element-path.js
Comment (jtnord) [inline core/src/main/resources/jenkins/formelementpath/form-element-path.js:178]: Ahhhh!!!!    I wonder if this may well explain some issues we have with codemirror sections not being found in the ATH!

**Review by jtnord — APPROVED** [review]
Thanks Tim, this looks like a great improvement.

Would like a review by @Vlatombe before this merges - as the enabling via system property may cause some issues for some of our testing.

File: core/src/main/java/jenkins/formelementpath/FormElementPathPageDecorator.java
Comment (Vlatombe) [reply core/src/main/java/jenkins/formelementpath/FormElementPathPageDecorator.java:13]: That's fine, we'll be able to provide the system property.

Comment (timja) [reply core/src/main/java/jenkins/formelementpath/FormElementPathPageDecorator.java:13]: Could also make it public as well

Comment (jtnord) [reply core/src/main/java/jenkins/formelementpath/FormElementPathPageDecorator.java:13]: as Vincent is ok with being able to provide it, we can keep it as it is for now (easier to open up, than restrict accesses later)

**Comment by timja** [comment]
@jglick your comments should be addressed now if you want to take another look

**Comment by timja** [comment]
or @Vlatombe you able to approve?

**Review by Vlatombe — APPROVED** [review]

**Comment by timja** [comment]
This PR is now ready for merge, after ~24 hours, we will merge it if there's no negative feedback.

Thanks!

**Comment by timja** [comment]
PR to form-element-path to kill it off: https://github.com/jenkinsci/form-element-path-plugin/pull/12

**Comment by timja** [comment]
Example using form element path to make a test more stable / easier to maintain:
https://github.com/jenkinsci/jenkins/pull/5923/commits/9281ae82c6ee43013cb4f226d6d57d1919053b16

This test broke every time markup was touched a little bit

File: core/src/main/java/jenkins/formelementpath/FormElementPathPageDecorator.java
Comment (jglick) [reply core/src/main/java/jenkins/formelementpath/FormElementPathPageDecorator.java:14]: That should set `isUnitTest` IIRC.

File: core/src/main/java/jenkins/formelementpath/FormElementPathPageDecorator.java
Comment (jglick) [reply core/src/main/java/jenkins/formelementpath/FormElementPathPageDecorator.java:13]: I think we need to turn this off by default, even in tests, at least until it can be rewritten to not rely on DOM listeners: https://github.com/jenkinsci/bom/pull/746#issuecomment-985172193

Comment (timja) [reply core/src/main/java/jenkins/formelementpath/FormElementPathPageDecorator.java:13]: Ah that’s a shames :(

Comment (jglick) [reply core/src/main/java/jenkins/formelementpath/FormElementPathPageDecorator.java:13]: ⇒ #6004

**Pull request merged**
